src/pint/models/noise_model.py

```diff
@@ -378,18 +397,23 @@ def ecorr_basis_weight_pair(self, toas):
         The weights used are the square of the ECORR values.
 
         """
+        no_toa_params = self.validate_toas(toas, suppress_warning=True)
         tbl = toas.table
         t = (tbl["tdbld"].quantity * u.day).to(u.s).value
         ecorrs = self.get_ecorrs()
         umats = []
         for ec in ecorrs:
+            if ec.name in no_toa_params:
+                continue
             mask = ec.select_toa_mask(toas)
             umats.append(create_quantization_matrix(t[mask]))
         nc = sum(u.shape[1] for u in umats)
         umat = np.zeros((len(t), nc))
         weight = np.zeros(nc)
         nctot = 0
         for ct, ec in enumerate(ecorrs):
+            if ec.name in no_toa_params:
+                continue
             mask = ec.select_toa_mask(toas)
```

Comment:
Select again gives you the right tool to skip the ones that aren't needed. And it's more efficient than searching through all the maskParameters in a separate loop.